ap/calibration/seeds.py

```diff
@@ -0,0 +1,112 @@
+# coding: utf-8
+
+"""
+Methods related to creating event and object seeds during calibration.
+"""
+
+import hashlib
+
+from ap.calibration import calibrator
+from ap.util import maybe_import, primes
+from ap.columnar_util import Route, get_ak_indices, set_ak_column, has_ak_column
+
+np = maybe_import("numpy")
+ak = maybe_import("awkward")
+
+
+@calibrator(
+    uses={
+        # global columns for event seed
+        "event", "nGenJet", "nGenPart", "nJet", "nPhoton", "nMuon", "nElectron", "nTau", "nSV",
+        # first-object columns for event seed
+        "Tau.jetIdx", "Tau.decayMode",
+        "Muon.jetIdx", "Muon.nStations",
+        "Jet.nConstituents", "Jet.nElectrons", "Jet.nMuons",
+    },
+    produces={"deterministic_seed"},
+)
+def deterministic_event_seeds(events, create_seed, **kwargs):
+    # start with a seed of the event number itself as the most sensitive integer
+    seed = create_seed(events.event)
+
+    # get global integers
+    global_fields = ["nGenJet", "nGenPart", "nJet", "nPhoton", "nMuon", "nElectron", "nTau", "nSV"]
+    for i, f in enumerate(global_fields, 3):
+        seed = seed + primes[i] * (events[f] if f in events.fields else ak.num(events[f[1:]]))
+
+    # get first-object integers
+    object_fields = [
+        "Tau.jetIdx", "Tau.decayMode", "Muon.jetIdx", "Muon.nStations", "Jet.nConstituents",
+        "Jet.nElectrons", "Jet.nMuons",
+    ]
+    for i, f in enumerate(object_fields, i + 1):
```

Comment:
It seems odd (and a little prone to errors) to me that the starting point of `enumerate` depends on the local variable of the loop in [line 33](https://github.com/uhh-cms/analysis_playground/pull/40/files#diff-34ad48c89852a085bc359ba5fdd0a38777a80f202215a459f1b0549b58d89f63R33). Is this on purpose? If yes, why not set the numbers separately, e.g.

```suggestion
# set the length separately
n_global_fields = len(global_fields)

# choose the same offset as in the event seed loop
for i, f in enumerate(object_fields, n_global_fields+3):
```

Reply:
Yeah, this was kinda of on purpose but I agree that it's not very clean to use the "leaking" index from the previous loop. I'll add the change you suggested (just with an inline call to `len`).